ev3dev2/motor.py

```diff
@@ -1924,6 +2037,181 @@ def get_speed_steering(self, steering, speed):
+
+        if distance_mm is not None and distance_mm < 0:
+            log.warning("{}: distance_mm is {}, it must be None or greater than 0".format(self, distance_mm))
+            abort = True
+
+        if abort:
+            self._set_brake(brake)
+            return
+
+        # The circle formed at the halfway point between the two wheels is the
+        # circle that must have a radius of radius_mm
+        circle_outer_mm = 2 * pi * (radius_mm + (self.wheel_distance_mm / 2))
+        circle_middle_mm = 2 * pi * radius_mm
+        circle_inner_mm = 2 * pi * (radius_mm - (self.wheel_distance_mm / 2))
+
+        if arc_right:
+            # The left wheel is making the larger circle and will move at 'speed'
+            # The right wheel is making a smaller circle so its speed will be a fraction of the left motor's speed
+            left_speed = speed
+            right_speed = float(circle_inner_mm/circle_outer_mm) * left_speed
+
+        else:
+            # The right wheel is making the larger circle and will move at 'speed'
+            # The left wheel is making a smaller circle so its speed will be a fraction of the right motor's speed
+            right_speed = speed
+            left_speed = float(circle_inner_mm/circle_outer_mm) * right_speed
+
+        log.debug("%s: arc %s, radius %s, distance %s, left-speed %s, right-speed %s, circle_outer_mm %s, circle_middle_mm %s, circle_inner_mm %s" %
+            (self, "right" if arc_right else "left",
+             radius_mm, distance_mm, left_speed, right_speed,
+             circle_outer_mm, circle_middle_mm, circle_inner_mm
+            )
+        )
+
+        if distance_mm is None:
+            MoveTank.on(self, left_speed, right_speed, brake, block)
+        else:
+            # We know we want the middle circle to be of length distance_mm so
+            # calculate the percentage of circle_middle_mm we must travel for the
+            # middle of the robot to travel distance_mm.
+            circle_middle_percentage = float(distance_mm / circle_middle_mm)
+
+            # Now multiple that percentage by circle_outer_mm to calculate how
+            # many mm the outer wheel should travel.
+            circle_outer_final_mm = circle_middle_percentage * circle_outer_mm
+
+            outer_wheel_rotations = float(circle_outer_final_mm / self.wheel.circumference_mm)
+            outer_wheel_degrees = outer_wheel_rotations * 360
+
+            log.debug("%s: arc %s, circle_middle_percentage %s, circle_outer_final_mm %s, outer_wheel_rotations %s, outer_wheel_degrees %s" %
+                (self, "right" if arc_right else "left",
+                 circle_middle_percentage, circle_outer_final_mm,
+                 outer_wheel_rotations, outer_wheel_degrees
+                )
+            )
+
+            MoveTank.on_for_degrees(self, left_speed, right_speed, outer_wheel_degrees, brake, block)
+
+    def on_arc_right(self, speed, radius_mm, distance_mm, brake=True, block=True):
```

Comment:
I'm having a bit of trouble understanding the use case for the API being structured like this (having separate "left" and "right" versions with radii), could you give some examples?

Reply:
What I'm trying to figure out is... what is the use case for being able to _arc with a radius_ for the purposes of driving? What cases do you know the radius and want to drive that circle?

Also, this brings up a related point: We have separate classes for "tank", "steering", and other drive patterns, but there isn't explicitly a reason that one needs to use _just tank_ or _just steering_. Certainly in some scenarios it makes sense, but if I'm trying to drive a particular series of lines and curves, it would be nice to be able to mix the various forms. I am pretty sure that works with our existing structure but it doesn't feel like it was designed to do so.

Reply:
I could see it being applicable in FLL where you want to drive in a partial circle to get around some object. Also the ability to drive in a circle of a specific radius was one of the main asks in the issue that was opened. 

I think MoveTank and MoveSteering are separate because EV3-G has these as separate blocks? Maybe we can combine these into MoveTank and have MoveSteering be a child class that doesn’t change anything? I need to look through exactly what MoveSteering does though. 

Heck could we also put all of this MoveDifferential code into MoveTank? Seems like we could